### ISSUE TITLE
Idea: Post-processing rules to handle analog needle obscuring digital registers 

### ISSUE THREAD

**Issue by b-neufeld** [jomjol/AI-on-the-edge-device#717]
Here's a picture of the meter I have in my home and ROIs configured. 

![image](https://user-images.githubusercontent.com/18536548/156930768-250e81ff-0d48-4894-9969-c55c06808f04.png)

The existing AI is quite good at reading the position of the analog register in the ROI I've configured. The challenge is that the analog meter with precision 0000.00**XX** will obscure parts the digital register (digits XXXX.XX), causing faulty readings on the digital registers.

For example if the analog needle reads ~0.0054, the needle will fall across the 4th digit in the digital counter. If dig4 is an 8, the AI may recognize it as a 3 (if you can imagine the needle covering the left side of an "8" digit). 

I've tried to accommodate this using the AllowNegativeRates, MaxRateValue, and MaxRateType and accepting some periodic errors, but I still end up getting "stuck" on something nonsensical and having to go reset the preValue to the current reading. 

My idea is to have some postprocessing logic that can be enabled in config.ini. The logic would be (in pseudocode): 
```
if limit_1 < ROIx < limit_2 then:
   use previous value for {list of ROIs} 
   use new/raw value for {list of ROIs} 
```

So for my meter I would configure something like this: 
```
if 3.5 < ana1 < 7.0 then:
   use previous value for {dig1, dig2, dig3, dig4, dig5, dig6} 
   use new/raw value for {null} 
```

Cheers!

**Comment by jomjol** [comment]
I thinkg, here you are at the limit. This kind of analog pointer is not similar to a teached one, therefore it will not work very well.
You would need to make trainings data for this.

**Comment by b-neufeld** [comment]
Thanks for the reply. I don't think the problem is training the needle - it's surprisingly good at reading an analog value from what I showed above. 

The problem is when this happens: 
![image](https://user-images.githubusercontent.com/18536548/156943982-5b9b7ea4-3601-44c1-950d-b82d24159dd9.png)

In this case, the 6th digital digit is currently unreadable. If the needle were a little further along as I've shown in MS Paint, it could turn an "8" into a "3" and cause a mis-read on that digit. 

Maybe there are two solutions to this. Extra post-processing logic, like I proposed above, or a set of training data with the entire face of the dial? 

Still, even if the analog dial was perfectly trained, if it obscures the digital registers, I won't be able to get a valid meter reading.

**Comment by jomjol** [comment]
Okay, now I got your point. This is a very specific issue, that would need a handling of the digits depending on the reading of the analog value. To be honest, that would be quite some effort, that I cannot invest currently. But if you adjust the post processing, feel free to do so!

**Comment by b-neufeld** [comment]
Thanks again for the reply! 

I will take a look at this myself. It won't be fast, but it will be something new for me to learn. I see the Build Instructions page has the basics to get me plugged into the project. 

Maybe just one more question, if you have time. Any clues or hints or places to get started? Just taking a quick look I am guessing somewhere in "flowcontrol", maybe ClassFlowPostProcessing.cpp, ClassFlowControll.cpp, and maybe ClassFlowCNNGeneral.cpp. Where is the function that assembles individual digits into a single number? 

Cheers!

**Comment by irakhlin** [comment]
I have had the same issue and I have handled it a little differently.
First I do not read the last digit from my meter and simply set in the configuration to shift the decimal, my water bill is charged per 100 cubit feet units so technically even the last digit isn't fully needed. Then I made sure to enable "allowed negative rates" to false, set the update interval to 4 minutes, and lastly I did a basic calculation, I have a 5/8 copper water meter, that has a max theoretical flow rate of 20 GPM, (4 x 20 = 80 gallons, which is 10.6944 cubit feet). This tells me I cant possibly have a larger change than 11 cubit feet per reading. I used this number as a starting point but basically ended up multiplying it by 4 ( say 3 readings in a row with actual change are bad) so I set MaxRateValue to 40. 

This isn't foolproof but thus far I have not had any "bad" readings make it to as a published message. It will simply delay and skip readings until the arrow gets to a position where its able to capture a valid reading.

![image](https://user-images.githubusercontent.com/12637936/158051927-935f3ab6-b79c-4311-aef0-d6469a767ae4.png)

**Comment by jomjol** [comment]
> Maybe just one more question, if you have time. Any clues or hints or places to get started? Just taking a quick look I am guessing somewhere in "flowcontrol", maybe ClassFlowPostProcessing.cpp, ClassFlowControll.cpp, and maybe ClassFlowCNNGeneral.cpp. Where is the function that assembles individual digits into a single number?

@b-neufeld The postprocessing is done in **ClassFlowPostProcessing**. There you have all digitized value available (comming from the ClassFlowCNNGeneral) and you can build the logic in there.

I will close this issue, to keep the overview. But feel free to contact me through this or any other issue.

**Comment by b-neufeld** [comment]
@jomjol and @irakhlin, thanks for the ideas and feedback. 

I decided not to attempt to edit ClassFlowPostProcessing... I just don't have the skills with the language and the dev environment. 

I came up with a pretty good band-aid solution that I'll share in case others need some inspiration - I'm not asking anyone to do any work for me! I admit this is a bit of an embarrassing, low-quality hack, but it should work :) 

I use Node-RED to grab values from the MQTT server, then push them into my influxdb historian, where I pull them into Home Assistant for automations and dashboards. I realized I could do this processing in Node-RED. If someone doesn't have influx they could buffer the last value in Node-RED and/or push the new values back to the MQTT server under a new topic. 

The flow looks like this: 

![image](https://user-images.githubusercontent.com/18536548/158475120-e4d6b327-30f3-4745-b452-a9033603a161.png)

Function code looks like this:
```
// Grab current reading from the meter
var current_reading = parseFloat(msg.payload["water_meter/main/json"].value);

// grab the last two digits of the current reading
var analog = ((current_reading * 100) - Math.floor(current_reading * 100)) * 100;

// grab the last GOOD value from the historian 
var last_value = parseFloat(msg.payload.last_value.last_value);

// last_value but only the digital registers (analog value stripped out)
var last_val_dig = Math.floor(last_value * 100)/100

// grab flow rate 
var flow_rate = parseFloat(msg.payload["water_meter/main/json"].rate);

// if analog needle is in the overlap zone, use the 6 most significant digits from the historian, and tack on the analog value
var new_reading = 0
// if new reading is NaN or new reading < old reading...
if (isNaN(current_reading) | current_reading < last_value) {
    new_reading = last_value;
} else if (30 < analog < 70 ) {
    new_reading = last_val_dig + analog/10000;
} else {
    new_reading = current_reading;
    }

// if new reading is NaN or new reading < old reading...
if (isNaN(current_reading) | current_reading < last_value) {
    new_reading = last_value;
}

// https://github.com/jomjol/AI-on-the-edge-device/issues/277
// 3 minutes is the sampling rate on AIOE 
// The rate is (current - last) / time
// So value = m^3/minute
// So multiply by 1000 for litres/minute
// Force flow rate to zero under certain conditions 
if (flow_rate === "" | isNaN(flow_rate) | flow_rate < 0 | last_value === current_reading) { 
    flow_rate=0;
}
flow_rate = flow_rate * 1000;

newMsg = { payload: { flow_rate: flow_rate, new_reading: new_reading, current_reading: current_reading, analog: analog, last_value: last_value, last_val_dig: last_val_dig }};
return newMsg;
```

Probably not perfect but we'll see how it works. ✌

**Comment by b-neufeld** [comment]
Some follow up details in case they are useful to anyone else: 

I went several months without updating the AIOE firmware and my Node-RED fix wasn't adequately reading values from the elster C700 meter. 

Recently I've upgraded everything to the most current version, including the models to the recommended class11 for digits and class100 for analog. I'm blown away by how these models have a near-flawless recognition rate of digits, even when partially obscured by the needle. And when the needle obscures a reading and generates an "N", it seems like the error correction algorithm has improved a lot. 

Nice work to all the contributors! :)
